drivers/counter/counter_nrfx_rtc.c

```diff
@@ -68,226 +74,428 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
-	struct counter_nrfx_data *dev_data = get_dev_data(dev);
+	__ASSERT_NO_MSG(get_dev_data(dev)->guard_period <
+			get_dev_data(dev)->top);
+	NRF_RTC_Type  *rtc = get_nrfx_config(dev)->rtc;
+	nrf_rtc_event_t evt;
+	u32_t prev_val;
+	u32_t top;
+	u32_t now;
+	u32_t diff;
+	u32_t int_mask = RTC_CHANNEL_INT_MASK(chan);
 	int err = 0;
-
-	for (int i = 0; i < counter_get_num_of_channels(dev); i++) {
-		/* Overflow can be changed only when all alarms are
-		 * disables.
-		 */
-		if (nrfx_config->ch_data[i].callback) {
-			return -EBUSY;
-		}
+	u32_t max_abs_val;
+	bool absolute = flags & COUNTER_ALARM_CFG_ABSOLUTE;
+	bool trigger_late = flags & COUNTER_ALARM_CFG_EXPIRE_WHEN_LATE;
+
+	__ASSERT(nrf_rtc_int_is_enabled(rtc, int_mask) == 0,
+			"Expected that CC interrupt is disabled.");
+
+	evt = RTC_CHANNEL_EVENT_ADDR(chan);
+	top =  get_dev_data(dev)->top;
+	max_abs_val = top - get_dev_data(dev)->guard_period;
+	now = nrf_rtc_counter_get(rtc);
+
+	/* First take care of a risk of an event coming from CC being set to
+	 * next tick. Reconfigure CC to future (now tick is the furtherest
+	 * future). If CC was set to next tick we need to wait for up to 15us
+	 * (half of 32k tick) and clean potential event. After that time there
+	 * is no risk of unwanted event.
+	 */
+	prev_val = nrf_rtc_cc_get(rtc, chan);
+	nrf_rtc_event_clear(rtc, evt);
+	nrf_rtc_cc_set(rtc, chan, now);
+	nrf_rtc_event_enable(rtc, int_mask);
+
+	if (ticks_sub(dev, prev_val, now, top) == 1) {
+		NRFX_DELAY_US(15);
+		nrf_rtc_event_clear(rtc, evt);
 	}
 
-	nrfx_rtc_cc_disable(rtc, TOP_CH);
 
-	dev_data->top_cb = cfg->callback;
-	dev_data->top_user_data = cfg->user_data;
-	dev_data->top = cfg->ticks;
-	nrfx_rtc_cc_set(rtc, TOP_CH, cfg->ticks, false);
+	/* Special handling required if value is next tick as that will not
+	 * generate event. Set COMPARE to <now> + 2. If meanwhile RTC counter
+	 * got incremented it means it's too late and alarm should expire.
+	 */
+	now = nrf_rtc_counter_get(rtc);
 
-	if (!(cfg->flags & COUNTER_TOP_CFG_DONT_RESET)) {
-		nrfx_rtc_counter_clear(rtc);
-	} else if (counter_nrfx_read(dev) >= cfg->ticks) {
-		err = -ETIME;
-		if (cfg->flags & COUNTER_TOP_CFG_RESET_WHEN_LATE) {
-			nrfx_rtc_counter_clear(rtc);
-		}
+	if (absolute) {
+		val = skip_zero_on_custom_top(val, top);
+	} else {
+		/* Half of top value is taken as an arbitrary value. It is
+		 * assumed that CC might be set too late only if requested
+		 * relative timeout is shorter than half of top value.
+		 */
+		trigger_late = val < (top / 2);
+		/* limit max to detect short relative being set too late. */
+		max_abs_val = trigger_late ? top / 2 : top;
+		val = ticks_add(dev, now, val, top);
 	}
 
-	if (cfg->callback || sw_wrap_required(dev)) {
-		nrfx_rtc_int_enable(rtc, COUNTER_TOP_INT_MASK);
+	diff = ticks_sub(dev, val, now, top);
+	if (diff == 1) {
+		val = ticks_add(dev, val, 1, top);
+		nrf_rtc_cc_set(rtc, chan, val);
+		get_dev_data(dev)->cc_adjusted |= BIT(chan);
```

Comment:
`cc_adjusted` should be also protected with atomic access, just like `cc_int_pending`.

Reply:
done